model/map-jpa/src/main/resources/META-INF/user-sessions/client/jpa-client-sessions-changelog-1.xml

```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright 2022 Red Hat, Inc. and/or its affiliates
+and other contributors as indicated by the @author tags.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd
+                    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <changeSet author="keycloak" id="client-sessions-1">
+
+        <createTable tableName="kc_client_session">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="version" type="INTEGER" defaultValueNumeric="0">
+                <constraints nullable="false"/>
+            </column>
+            <column name="usersessionid" type="UUID">
+                <constraints foreignKeyName="client_session_fk_user_session_fkey" references="kc_user_session(id)" deleteCascade="true" deferrable="true" initiallyDeferred="true" />
```

Comment:
The deferrable/initiallyDeferring don't work on CockroachDB AFAIK, therefore I suggest to put this into a separate changeset that only runs for PostgreSQL. For CockroachDB, there should be an index on that column instead to be able to efficiently query all client sessions for a user session. 

To make the cleanup  work on CockroachDB, I suppose there needs to be a change in this issue to delete all client session when a user session is removed (critieria query)?, and a new issue to be implemented later with a housekeeping mechanism that cleans up client_session that might be left-over. Until that is implemented we could count on the regular expiry of client sessions.

Reply:
Thanks @ahus!

> The deferrable/initiallyDeferring don't work on CockroachDB AFAIK, therefore I suggest to put this into a separate changeset that only runs for PostgreSQL. 

You're right deferrable/initiallyDeferring don't work on CockroachDB, at least not yet. For the reference there is opened issue for that: https://go.crdb.dev/issue-v/31632/v21.2

> For CockroachDB, there should be an index on that column instead to be able to efficiently query all client sessions for a user session.

There is index for that column with name `client_session_user_session_id_fk` for both PostgreSQL and CockroachDB

> To make the cleanup work on CockroachDB, I suppose there needs to be a change in this issue to delete all client session when a user session is removed (critieria query?), and a new issue to be implemented later with a housekeeping mechanism that cleans up client_session that might be left-over. Until that is implemented we could count on the regular expiry of client sessions.

I'd like to avoid adding any additional code for removing clients sessions just for CRDB. When we move the constraint into separate changeset which would be run for Postgres, in CRDB we'd leave client session stored without its user/session. Once we'll have the housekeeping mechanism to remove expired entities we might want to extend it to remove also orphaned client session. So the question is if we are ok with leaving some abandoned client session in CRDB for now. 

> UPDATE: maybe the MapUserSessionProvider.removeUserSession(realm, session)/removeUserSessions(realm, user) could call also call clientSessionTx.delete() ... it might be less straightforward for removeUserSessions(realm).
It might be reasonable to do some things in a separate issue as long as it works with CockroachDB now.

I was proposing something like that within https://github.com/keycloak/keycloak/issues/12096#issuecomment-1136921712 and it turned out we need to let each implementation provide its own way how to implement this cascade removal. While on PostgreSQL it works fine, on CRDB unfortunately not.

Reply:
Oh, I forgot that the code in MapUserSessionProvider is executed for all store implementations. Hm, ok. Maybe it is ok to keep it as is for now. 

Then I'd advocate to only the foreign key to a PostgreSQL changeset. And if it is re-modeled to remove the `kc_user_session_client_session`, it might even work with Hibernate's orphan removal?! Let's see...